### PR TITLE
Split Rank1 MdArray allocation into a separate helper

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
MdArray rank1 with all-zero lower bounds gets actually allocated as an SzArray. On Native AOT this means we need to spin up to the type loader to potentially load the new type. Split this expensive (and impossible-to-express-in-C#) thing into a separate helper so we can trim the type loader.

https://github.com/dotnet/runtime/pull/84156#issuecomment-1491402863

Cc @dotnet/ilc-contrib

### CODE REVIEW

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:58]: The `(eeType.IsSzArray)` from above should be moved to the other helper as well.

File: src/coreclr/tools/Common/Internal/Runtime/ReadyToRunConstants.cs
Comment (jkotas) [inline src/coreclr/tools/Common/Internal/Runtime/ReadyToRunConstants.cs:355]: Adding this helper here would require R2R version bump.

Comment (MichalStrehovsky) [reply src/coreclr/tools/Common/Internal/Runtime/ReadyToRunConstants.cs:355]: It comes after FirstFakeHelper - where do those show up in the file format?

Comment (jkotas) [reply src/coreclr/tools/Common/Internal/Runtime/ReadyToRunConstants.cs:355]: I have missed that. Ignore this comment.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:58]: Actually, the condition in the JIT does not look right either. It will make us use the special helper for the jagged arrays - we do not want that.

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:58]: I do not see a method on JIT/EE interface that can be used to distinguish 2-dimensional jagged array vs. 1-dimensional md-array with two args. We may need a new method on the JIT/EE interface for this.

Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:58]: Can the jagged arrays be constructed with regular C#? We could just call into the old helper to handle the rest of cases. It's really just about splitting off the edge cases from the mainstream thing.

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:58]: Good point. It makes sense to have a helper that handles just what is used by C# - true MD arrays of rank 2 or more, and second helper for everything else.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:57]: We are handling IsSzArray case a few lines below. Also, the recursive call in the IsSzArray case handling needs to call the Rare overload.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:27]:
```suggestion
            // Rank 1 arrays are handled below.
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:82]:
```suggestion
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:80]:
```suggestion
                const int rank = 1;
```

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ArrayHelpers.cs:60]:
```suggestion
            Debug.Assert(nDimensions > 0);

            Debug.Assert(eeType.ArrayRank == 1);

            if (eeType.IsSzArray)
```

**Review by jkotas — COMMENTED** [review]
Some nits

**Review by jkotas — APPROVED** [review]
Thanks!

**Comment by MichalStrehovsky** [comment]
Cc @dotnet/jit-contrib since I'm touching the JIT. Also I know there's extra logistics to think about when JitInterface GUID is changed: I don't mind this waiting until a merge window opens.

**Review by EgorBo — APPROVED** [review]

**Comment by EgorBo** [comment]
> Cc @dotnet/jit-contrib since I'm touching the JIT. Also I know there's extra logistics to think about when JitInterface GUID is changed: I don't mind this waiting until a merge window opens.

I don't think we have a strict rule around that. I personally try to merge such changes on weekends since SPMI re-collection is a bit fragile (takes a lot of time, some collections aren't automated yet), if @AndyAyersMS hasn't yet updated aspnet I assume we can merge this one today

**Comment by AndyAyersMS** [comment]
> if @AndyAyersMS hasn't yet updated aspnet I assume we can merge this one today

Merging today is fine by me.

**Pull request merged**
